core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java

```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
         return null;
     }
 
+    private void interruptRunawaySplits()
+    {
+        for (RunningSplitInfo splitInfo : runningSplitInfos) {
+            Duration duration = Duration.succinctNanos(ticker.read() - splitInfo.getStartTime());
+
+            if (duration.compareTo(interruptRunawaySplitsTimeout) > 0) {
+                List<StackTraceElement> stackTraceElements = asList(splitInfo.getThread().getStackTrace());
+
+                if (!splitInfo.printed) {
+                    splitInfo.setPrinted();
+
+                    String splitThreadStack = toStackString(stackTraceElements);
+                    String splitThreadId = splitInfo.getThreadId();
+                    log.warn(String.format("Split %s has been long running %n%s", splitInfo.getSplitInfo(), splitThreadStack));
+                }
+
+                if (interruptibleSplitPredicate.test(stackTraceElements)) {
+                    log.warn("Potentially interrupting runaway split matching the predicate " + splitInfo.getSplitInfo());
+
+                    synchronized (splitInfo.getTaskRunner()) {
```

Comment:
Instead of synchronizing on the task runner here I would recommend having a `splitInfo.getTaskRunner().interruptSplit(splitInfo)` to have the synchronization well encapsulated

Reply:
done.